### PR TITLE
feat(helper): streaming-sse

### PR DESCRIPTION
**Pull request by watany-dev**
As I work on supporting SSE for the Lambda Adaptor, I am creating a general-purpose helper. The reason the test code is inadequate is because of the lock error I'm encountering. If you're open to embracing this helper, I would greatly appreciate any insights or ideas you might have on this.

```bash
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Errors ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

Vitest caught 1 unhandled error during the test run.
This might cause false positive tests. Resolve unhandled errors to make sure your tests are not affected.

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Unhandled Rejection ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
TypeError: Invalid state: WritableStream is locked
 ❯ new NodeError node:internal/errors:405:5
 ❯ setupWritableStreamDefaultWriter node:internal/webstreams/writablestream:594:11
 ❯ new WritableStreamDefaultWriter node:internal/webstreams/writablestream:402:5
 ❯ WritableStream.getWriter node:internal/webstreams/writablestream:264:12
 ❯ new StreamingApi src/utils/stream.ts:8:28
      6|   constructor(writable: WritableStream) {
      7|     this.writable = writable
      8|     this.writer = writable.getWriter()
       |                            ^
      9|     this.encoder = new TextEncoder()
     10|   }
 ❯ new SSEStreamingApi src/helper/stream/index.ts:12:5
 ❯ src/helper/stream/index.ts:38:20
 ❯ Context.stream src/context.ts:396:5
 ❯ Module.streamAsSSE src/helper/stream/index.ts:37:12
 ❯ src/helper/stream/index.test.ts:10:17

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Serialized Error: { code: 'ERR_INVALID_STATE' }
This error originated in "src/helper/stream/index.test.ts" test file. It doesn't mean the error was thrown inside the file itself, but while it was running.
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 Test Files  52 passed (52)
      Tests  815 passed (815)
     Errors  1 error
   Start at  09:58:26
   Duration  22.17s (transform 7.42s, setup 16ms, collect 23.54s, tests 16.66s, environment 8.90s, prepare 37.28s)
```

### Author should do the followings, if applicable

- [x] Add tests
- [ ] Run tests
- [ ] `yarn denoify` to generate files for Deno

### CODE REVIEW

File: src/helper.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // This file is for Deno to import helpers from `hono/helper.ts`.
+export * from './helper/adapter'
 export * from './helper/cookie'
 export * from './helper/html'
-export * from './helper/adapter'
+export * from './helper/stream'
```
Comment (yusukebe) [inline src/helper.ts:5]: `./helper/streaming/index.ts` should be better.

File: src/helper/stream/index.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        `data: ${message.data}`,
+      ]
+        .filter(Boolean)
+        .join('\n') + '\n'
+
+    await this.write(sseData)
+  }
+}
+
+function setSSEHeaders(context: Context) {
```
Comment (yusukebe) [inline src/helper/stream/index.ts:29]: Use the following syntax to be uniform with others.

```ts
const setSSEHeaders = () => {}
```

File: src/helper/stream/index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  context.set('Transfer-Encoding', 'chunked')
+  context.set('Content-Type', 'text/event-stream')
+  context.set('Cache-Control', 'no-cache')
+  context.set('Connection', 'keep-alive')
+}
+
+export const streamAsSSE = (c: Context, cb: (stream: SSEStreamingApi) => Promise<void>) => {
```
Comment (yusukebe) [inline src/helper/stream/index.ts:36]: I think `streamSSE` is better and simple.

File: src/helper/stream/index.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  event?: string
+  id?: string
+}
+
+class SSEStreamingApi extends StreamingApi {
+  constructor(originalStream: StreamingApi) {
+    super(originalStream['writable']) // Accessing the private writable property - normally not recommended but works for this specific solution.
+  }
+
+  async writeSse(message: SSEMessage) {
```
Comment (yusukebe) [inline src/helper/stream/index.ts:15]: You can write `writeSse` and `writeSSE`, what do you think? I think `writeSSE` is better to unify with others.

**Comment by yusukebe** [comment]
Hi @watany-dev !

I have commented on a few minor points, but I would like to accept this PR! Could you move on to the next action?

**Comment by watany-dev** [comment]
@sor4chi 

Do you know anything about this write lock issue?

**Comment by sor4chi** [comment]
Hi, @watany-dev. Thank you for implementing this helper!

I think this is because when you set up an instance of `StreamingAPI`, it is designed to get `writer` from the constructor argument `writable`, so you cannot reuse `StreamingAPI.writable`, which already gets `writer`, as the constructor argument of `StreamingAPI`.

How about this kind of implementation?
The approach is to create another `reader`/`writer` pair and pipe it to the `originalStream`.

```diff
diff --git a/src/helper/stream/index.ts b/src/helper/stream/index.ts
index f4a28ae..ff94178 100644
--- a/src/helper/stream/index.ts
+++ b/src/helper/stream/index.ts
@@ -8,8 +8,8 @@ interface SSEMessage {
 }
 
 class SSEStreamingApi extends StreamingApi {
-  constructor(originalStream: StreamingApi) {
-    super(originalStream['writable']) // Accessing the private writable property - normally not recommended but works for this specific solution.
+  constructor(writable: WritableStream) {
+    super(writable)
   }
 
   async writeSse(message: SSEMessage) {
@@ -35,7 +35,9 @@ function setSSEHeaders(context: Context) {
 
 export const streamAsSSE = (c: Context, cb: (stream: SSEStreamingApi) => Promise<void>) => {
   return c.stream(async (originalStream: StreamingApi) => {
-    const stream = new SSEStreamingApi(originalStream)
+    const { readable, writable } = new TransformStream()
+    const stream = new SSEStreamingApi(writable)
+    originalStream.pipe(readable)
     setSSEHeaders(c)
 
     try {
```

File: src/helper.ts (hunk 1 above)
Comment (watany-dev) [reply src/helper.ts]: fixed

File: src/helper/stream/index.ts (hunk 2 above)
Comment (watany-dev) [reply src/helper/stream/index.ts]: fixed!

File: src/helper/stream/index.ts (hunk 3 above)
Comment (watany-dev) [reply src/helper/stream/index.ts]: fixed.

File: src/helper/stream/index.ts (hunk 4 above)
Comment (watany-dev) [reply src/helper/stream/index.ts]: fixed

**Comment by watany-dev** [comment]
@sor4chi 
I'm glad to hear that your project is working smoothly with the help of the idea. I'd be happy to take a look at the changes you've made. 

@yusukebe 
Please go ahead and share the details or any specific code you'd like me to review.

**Comment by watany-dev** [comment]
@sor4chi 
Thank you. I am grateful that it now works smoothly thanks to your wonderful idea. 

@yusukebe 
I've incorporated your idea. Could you please take a look?

**Review by yusukebe — APPROVED** [review]
LGTM!

**Comment by yusukebe** [comment]
@watany-dev 

I'll merge this PR. It will be included in the next minor version update. I will create a `next` branch for the version up in a while and merge it there.

Thanks!

**Pull request merged**